R/adjust-numeric-calibration.R

```diff
@@ -0,0 +1,74 @@
+#' Re-calibrate numeric predictions
+#'
+#' @param x A [container()].
+#' @param calibrator A pre-trained calibration method from the \pkg{probably}
+#' package, such as [probably::cal_estimate_linear()].
+#' @export
+adjust_numeric_calibration <- function(x, calibrator) {
```

Comment:
Currently, this function (and it's probability analog) take a trained `calibrator` and then don't do any actual fitting at `fit()` time. This means that:

* If users are not using a workflow, they need to specify the names of `truth` and `estimate` (and by different names!).
*  If they're using a workflow, they now need to locate the names of `truth` and `estimate` when they didn't used to need to.
* Users (and tune, downstream) need to train the calibrator outside of the workflow. 

With the current draft, this looks like:

``` r
library(modeldata)
library(probably)
#> 
#> Attaching package: 'probably'
#> The following objects are masked from 'package:base':
#> 
#>     as.factor, as.ordered
library(tibble)
library(container)

# create example data
set.seed(1)
dat <- tibble(y = rnorm(100), y_pred = y/2 + rnorm(100))

dat
#> # A tibble: 100 × 2
#>         y y_pred
#>     <dbl>  <dbl>
#>  1 -0.626 -0.934
#>  2  0.184  0.134
#>  3 -0.836 -1.33 
#>  4  1.60   0.956
#>  5  0.330 -0.490
#>  6 -0.820  1.36 
#>  7  0.487  0.960
#>  8  0.738  1.28 
#>  9  0.576  0.672
#> 10 -0.305  1.53 
#> # ℹ 90 more rows

# calibrate numeric predictions
reg_cal <- cal_estimate_linear(dat, truth = y, estimate = y_pred)

# specify calibration 
reg_ctr <-
  container(mode = "regression") %>%
  adjust_numeric_calibration(reg_cal)

# "train" container
reg_ctr_trained <- fit(reg_ctr, dat, outcome = y, estimate = y_pred)

predict(reg_ctr, dat)
#> # A tibble: 100 × 2
#>         y  y_pred
#>     <dbl>   <dbl>
#>  1 -0.626 -0.257 
#>  2  0.184  0.303 
#>  3 -0.836 -0.512 
#>  4  1.60   0.479 
#>  5  0.330  0.0108
#>  6 -0.820  0.523 
#>  7  0.487  0.480 
#>  8  0.738  0.516 
#>  9  0.576  0.438 
#> 10 -0.305  0.536 
#> # ℹ 90 more rows
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

I propose that we supply some interface in place of `calibrator` that just specifies the kind of calibration that would happen, e.g. `type = "linear", ...` or `fn = cal_estimate_linear, ...` and actually fit the calibrator at `fit()` time. This would allow for workflows users to never have to specify column names and container users to only specify column names once. This would look something like:

``` r
library(modeldata)
library(probably)
#> 
#> Attaching package: 'probably'
#> The following objects are masked from 'package:base':
#> 
#>     as.factor, as.ordered
library(tibble)
library(container)

# create example data
set.seed(1)
dat <- tibble(y = rnorm(100), y_pred = y/2 + rnorm(100))

dat
#> # A tibble: 100 × 2
#>         y y_pred
#>     <dbl>  <dbl>
#>  1 -0.626 -0.934
#>  2  0.184  0.134
#>  3 -0.836 -1.33 
#>  4  1.60   0.956
#>  5  0.330 -0.490
#>  6 -0.820  1.36 
#>  7  0.487  0.960
#>  8  0.738  1.28 
#>  9  0.576  0.672
#> 10 -0.305  1.53 
#> # ℹ 90 more rows

# specify calibration (no training has happened yet)
reg_ctr <-
  container(mode = "regression") %>%
  adjust_numeric_calibration(type = "linear") # or `fn = cal_estimate_linear`, etc

# actually train the calibration in the container
reg_ctr_trained <- fit(reg_ctr, dat, outcome = y, estimate = y_pred)

predict(reg_ctr, dat)
#> # A tibble: 100 × 2
#>         y  y_pred
#>     <dbl>   <dbl>
#>  1 -0.626 -0.257 
#>  2  0.184  0.303 
#>  3 -0.836 -0.512 
#>  4  1.60   0.479 
#>  5  0.330  0.0108
#>  6 -0.820  0.523 
#>  7  0.487  0.480 
#>  8  0.738  0.516 
#>  9  0.576  0.438 
#> 10 -0.305  0.536 
#> # ℹ 90 more rows
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Reply:
The problem is: what data do you have to do that? 

In the course of tuning, you only have the holdout predictions for that resample. If we had the whole set of out-of-sample predictions, we could train on that. Using re-predicted training data isn't a good option. 

Another idea is to supply a `data` argument where people can pass in any predictions that they want. However, unless they have a specific validation set held out for this specific operation, we are back to creating those predictions outside of the workflow/tune process. 

I'm not happy with what the process is now but I think it is what we can do (for now, at least).

Reply:
I think that I'll write a blog post about this to outline my thoughts better (and we can refer to that).

Reply:
> The problem is: what data do you have to do that?

The current approach leans on users to figure that out--an approach like the one I've proposed allows us to take care of that for the user.

I think framing this in the context of tuning/resampling, like you've done, is helpful. When I think through the interface as implemented, I'm unsure how to imagine tuning parameters from the container. `tune_*()` takes in _one_ workflow and resamples it. In the current approach, how is that workflow's postprocessor trained? The calibrator needs predictions, and based on your reply, we don't want those predictions to come from data that the model was trained on. So, do we take a three-way split, train/test the validation split, train some model on the training portion of the validation split, train the calibrator on the testing portion of the validation split, and then pass that fixed calibrator to the workflow? When tuning, which of the resampled models is the one used to train the calibrator? e.g., as I understand it, this would be the resampling flow as implemented:

```r
library(tidymodels)
library(container)
library(probably)

penguins

# usual three-way flow
peng_split <- initial_validation_split(penguins)
peng_train <- training(peng_split)
peng_test <- testing(peng_split)
peng_val <- validation(peng_split)
peng_res <- vfold_cv(peng_train)

bt <- boost_tree("regression", trees = 3, learn_rate = tune())

# don't want to train the calibrator on re-predicted values
peng_split_cal <- initial_split(peng_val)
peng_train_cal <- training(peng_split_cal)
peng_test_cal <- testing(peng_split_cal)

# which of the models to be proposed trains the calibrator?
# bt_finalized <- finalize(bt, ...)
bt_cal_train <- fit(bt_finalized, body_mass_g ~ ., peng_train_cal)
bt_cal_preds <- augment(bt_cal_train, peng_test_cal)
bt_cal <- cal_estimate_linear(bt_cal_preds, ...)

bt_container <- container() %>% adjust_numeric_calibration(bt_cal)

wf <- workflow(body_mass_g ~ ., bt, bt_container)

tune_grid(wf, peng_res)
```

If we just allow specifying the strategy used to train a post-processor rather than the trained post-processor, the post-processor is actually resampled and the user interface feels much more like tidymodels. Also, the onus of doing the right thing w.r.t. resampling is on us rather than the user:

```r
library(tidymodels)
library(container)
library(probably)

penguins

# usual two-way flow
peng_split <- initial_split(penguins)
peng_train <- training(peng_split)
peng_test <- testing(peng_split)
peng_res <- vfold_cv(peng_train)

bt <- boost_tree("regression", trees = 3, learn_rate = tune())
bt_container <- container() %>% adjust_numeric_calibration(type = "linear")

wf <- workflow(body_mass_g ~ ., bt, bt_container)

tune_grid(wf, peng_res)
```

> In the course of tuning, you only have the holdout predictions for that resample. If we had the whole set of out-of-sample predictions, we could train on that. Using re-predicted training data isn't a good option.

From my understanding of our conversation earlier in the week, you've observed that training calibrators on re-predicted data is typically not a problem practically.

That said, from a technical standpoint, this feels quite solvable in tune:

1) **Take a computational performance penalty**: figure out a reasonably memory-efficient way to train calibrators on unseen data by exchanging out-of-sample predictions after the initial tune grid loop.
2) **Take a predictive performance penalty**: take an "internal" train/test split of the analysis (training portion of the resample) set for each resample when tuning post-processors and train the model on the internal training portion, post-processor on the internal testing portion.